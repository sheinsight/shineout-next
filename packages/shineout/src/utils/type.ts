export type RequireAWithB<T, A extends keyof T, B extends keyof T> = 
  | (Omit<T, A> & { [K in A]?: never })
  | (T & Required<Pick<T, B>>)
  | (T & Required<Pick<T, A | B>>)