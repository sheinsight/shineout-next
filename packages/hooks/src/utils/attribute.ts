export const getDataAttribute = (attrs: Record<string, string>) => {
  return Object.keys(attrs).reduce((acc, key) => {
    const ns = `data-soui-${key}`;
    acc[ns] = attrs[key];
    return acc;
  }, {} as Record<string, string>);
};
