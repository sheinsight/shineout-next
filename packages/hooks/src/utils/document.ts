import { isBrowser } from './is';
export const docSize = {
  get width() {
    if (!isBrowser()) return 0;
    return document.documentElement.clientWidth || document.body.clientWidth;
  },
  get height() {
    if (!isBrowser()) return 0;
    return document.documentElement.clientHeight || document.body.clientHeight;
  },
};
export const docScroll = {
  get top() {
    if (!isBrowser()) return 0;
    return document.documentElement.scrollTop || document.body.scrollTop;
  },
  get left() {
    if (!isBrowser()) return 0;
    return document.documentElement.scrollLeft || document.body.scrollLeft;
  },
  set top(value) {
    if (!isBrowser()) return;
    document.documentElement.scrollTop = value;
    document.body.scrollTop = value;
  },
  set left(value) {
    if (!isBrowser()) return;
    document.documentElement.scrollLeft = value;
    document.body.scrollLeft = value;
  },
};

export function isInDocument(element: HTMLElement | null) {
  if (element && 'isConnected' in element) {
    return element.isConnected;
  }
  return isBrowser() && document.documentElement.contains(element);
}
