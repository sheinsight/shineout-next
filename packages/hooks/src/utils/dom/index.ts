export { default, default as NormalizeWheel } from './normalize-wheel';
export type { OldWheelEvent } from './normalize-wheel.type';
