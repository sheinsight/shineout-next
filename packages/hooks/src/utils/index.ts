export * from './is';
export { default as extractEventHandlers } from './extractEventHandlers';
export * from './clone';
export * from './object';
export * from './string';
export * from './func';
export * from './validate';
export * from './number';
export * from './shallowEqual';
export * from './uid';
export * from './document';
export * from './position';
export * from './render';
export * from './lazyload';
export * from './react-render';
export * from './style-unit';
export * from './attribute';
