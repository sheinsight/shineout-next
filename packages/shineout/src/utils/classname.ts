import classnames from 'classnames';
import { config } from '@sheinx/base';

interface CSSModule {
  [key: string]: string;
}

/**
 * create a new className generate function, add namespace, handle css module
 * @param style - object; for css module
 * @param module - string
 * @param prefix - string, default value is 'shineout'
 * * */
export default (style: CSSModule, module: string, prefix: string = config.prefix) =>
  (...args: any) => {
    const className = classnames(...args);
    if (!className) return '';

    const ns = `${prefix}${module ? `-${module}` : '-'}`;
    let list: (string | number)[] = className
      .split(' ')
      .map((c) => (c === '_' ? ns : `${ns}-${c}`));
    if ((config as any).cssModule) {
      list = list.map((c: any) => style[c] || c);
    }
    return list.join(' ');
  };
