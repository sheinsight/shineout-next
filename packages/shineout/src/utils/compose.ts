export function compose(...funcs: any) {
  if (funcs.length === 0) {
    return (arg: any) => arg;
  }
  const last = funcs[funcs.length - 1];
  const rest = funcs.slice(0, -1);
  return (...args: typeof last) =>
    rest.reduceRight((composed: any, f: any) => f(composed), last(...args));
}
