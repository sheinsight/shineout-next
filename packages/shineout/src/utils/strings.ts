import { util } from "@sheinx/hooks";

const { devUseWarning } = util;

export function capitalize(str: string) {
  if (typeof str !== 'string') {
    devUseWarning.error('str should be a string');
  }
  return str && str[0].toUpperCase() + str.slice(1);
}

export function substitute(str: string | Function, obj: { [x: string]: any }) {
  if (typeof str === 'string') {
    if (str.indexOf('{') < 0) {
      return str;
    }

    return str.replace(/\\?\{([^{}]+)\}/g, (match, name) => {
      if (match.charAt(0) === '\\') {
        return match.slice(1);
      }
      return obj[name] === null || obj[name] === undefined ? '' : obj[name];
    });
  }
  if (typeof str === 'function') {
    let val = str(obj);
    if (val === obj && typeof val === 'object') {
      val = Object.assign({}, obj);
    }
    return val;
  }

  return '';
}

export function removeProtocol(url: string) {
  if (url.indexOf('http') !== 0) return url;
  try {
    const { href, protocol } = new URL(url);
    return href.slice(protocol.length);
  } catch (error) {
    return url;
  }
}

export function getRTLPosition(position: string) {
  if (!position) return position;
  // position.replace('left', 'right').replace('right', 'left')
  if (position.indexOf('left') !== -1) {
    return position.replace('left', 'right');
  }
  if (position.indexOf('right') !== -1) {
    return position.replace('right', 'left');
  }
  return position;
}

export function getDirectionIconName(mode = 'left', double = false) {
  const rtl = true;
  if (mode === 'left') {
    if (rtl) {
      return double ? 'ArrowRightDouble' : 'AngleRight';
    }
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }

  if (rtl) {
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }
  return double ? 'ArrowRightDouble' : 'AngleRight';
}
