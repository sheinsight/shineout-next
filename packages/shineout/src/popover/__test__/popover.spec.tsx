import { render, cleanup, screen, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Popover from '..';
import Button from '../../button';
import mountTest from '../../tests/mountTest';
import { attributesTest, classTest, delay, displayTest, styleTest } from '../../tests/utils';
import PopoverBase from '../__example__/01-base';
import PopoverControll from '../__example__/04-controll';

const SO_PREFIX = 'popover';
const popoverClassName = `.${SO_PREFIX}-wrapper-0-2-65`;
const popoverBaseClassName = `.${SO_PREFIX}-wrapper-0-2-1`;
const popoverContentClassName = `.${SO_PREFIX}-content-0-2-68`;
const popoverContentBaseClassName = `.${SO_PREFIX}-content-0-2-4`;
const popoverOpenClassName = `${SO_PREFIX}-wrapperOpen-0-2-66`;
const popoverOpenBaseClassName = `${SO_PREFIX}-wrapperOpen-0-2-2`;
const popoverArrowClassName = `.${SO_PREFIX}-arrow-0-2-3`;
const popoverArrowBaseClassName = `.${SO_PREFIX}-arrow-0-2-67`;
const popoverTextClassName = `${SO_PREFIX}-text-0-2-69`;
const popoverBaseTextClassName = `${SO_PREFIX}-text-0-2-5`;

// const positions = ['bottom-left', 'bottom', 'bottom-right', 'right-top', 'left-top', 'right', 'left', 'right-bottom', 'left-bottom', 'top-left', 'top', 'top-right']

beforeAll(() => {
  jest.useFakeTimers();
});
afterAll(() => {
  jest.runAllTimers();
});
afterEach(cleanup);
mountTest(<Popover />);

const PopoverDemo = ({ className, style, trigger, position }: any) => {
  return (
    <Button>
      Hover
      <Popover className={className} style={style} trigger={trigger} position={position}>
        some Text
      </Popover>
    </Button>
  );
};

const getPopoverRoot = () =>
  document.querySelector(popoverClassName)! || document.querySelector(popoverBaseClassName)!;
const getPopoverContent = () =>
  document.querySelector(popoverContentClassName)! ||
  document.querySelector(popoverContentBaseClassName)!;

describe('Popover[Snapshot]', () => {
  test('render snapshot default', async () => {
    const { container } = render(<PopoverBase />);
    fireEvent.mouseEnter(container.querySelector('button')!);
    await waitFor(async () => {
      await delay(200);
      expect(document.querySelector(popoverClassName)).toMatchSnapshot();
    });
  });
});

describe('Popover[Base]', () => {
  displayTest(Popover, 'ShineoutPopover');
  test('should render default', async () => {
    const { container } = render(<PopoverDemo />);
    let popover = getPopoverRoot();
    expect(popover).not.toBeInTheDocument();
    fireEvent.mouseEnter(container.querySelector('button')!);
    await waitFor(async () => {
      await delay(200);
      popover = getPopoverRoot();
      expect(
        popover.classList.contains(popoverOpenClassName) ||
          popover.classList.contains(popoverOpenBaseClassName),
      ).toBeTruthy();
      styleTest(
        popover,
        'position: absolute; z-index: 1060; left: 0px; transform: translateX(-50%); top: 0px;',
      );
      attributesTest(popover, 'data-soui-position', 'bottom');
      expect(
        popover.querySelector(popoverArrowClassName) ||
          popover.querySelector(popoverArrowBaseClassName),
      ).toBeTruthy();
      const content = getPopoverContent();
      expect(content.textContent).toBe('some Text');
      expect(
        content.classList.contains(popoverTextClassName) ||
          content.classList.contains(popoverBaseTextClassName),
      ).toBeTruthy();
    });
  });
  test('should render when set className and style', async () => {
    const { container } = render(<PopoverDemo className='demo' style={{ color: 'black' }} />);
    fireEvent.mouseEnter(container.querySelector('button')!);
    await waitFor(async () => {
      await delay(200);
      const popover = getPopoverRoot();
      classTest(popover, 'demo');
      const content = getPopoverContent();
      styleTest(content, 'color: black;');
    });
  });
  test('should render when set trigger is click', async () => {
    const { container } = render(<PopoverDemo trigger='click' />);
    expect(getPopoverRoot()).not.toBeInTheDocument();
    const button = container.querySelector('button')!;
    fireEvent.mouseEnter(button);
    await waitFor(async () => {
      await delay(200);
      expect(getPopoverRoot()).not.toBeInTheDocument();
    });
    fireEvent.click(button);
    await waitFor(async () => {
      await delay(200);
      expect(getPopoverRoot()).toBeInTheDocument();
    });
  });

  const positionStyleMap: Record<string, string> = {
    'bottom-left': 'position: absolute; z-index: 1060; left: 0px; top: 0px;',
    bottom: 'position: absolute; z-index: 1060; left: 0px; transform: translateX(-50%); top: 0px;',
    'bottom-right':
      'position: absolute; z-index: 1060; left: 0px; transform: translateX(-100%); top: 0px;',
    'right-top': 'position: absolute; z-index: 1060; top: 0px; left: 0px;',
    'left-top':
      'position: absolute; z-index: 1060; top: 0px; transform: translateX(-100%); left: 0px;',
    right: 'position: absolute; z-index: 1060; top: 0px; transform: translateY(-50%); left: 0px;',
    left: 'position: absolute; z-index: 1060; top: 0px; transform: translateY(-50%) translateX(-100%); left: 0px;',
    'right-bottom':
      'position: absolute; z-index: 1060; top: 0px; transform: translateY(-100%); left: 0px;',
    'left-bottom':
      'position: absolute; z-index: 1060; top: 0px; transform: translateY(-100%) translateX(-100%); left: 0px;',
    'top-left':
      'position: absolute; z-index: 1060; left: 0px; transform: translateY(-100%); top: 0px;',
    top: 'position: absolute; z-index: 1060; left: 0px; transform: translateX(-50%)translateY(-100%); top: 0px;',
    'top-right':
      'position: absolute; z-index: 1060; left: 0px; transform: translateX(-100%)translateY(-100%); top: 0px;',
  };

  Object.keys(positionStyleMap).forEach((item: string) => {
    test(`should render when set position is ${item}`, async () => {
      const { container } = render(<PopoverDemo position={item} trigger='click' />);
      fireEvent.click(container.querySelector('button')!);
      await waitFor(async () => {
        await delay(200);
        const popover = getPopoverRoot();
        attributesTest(popover, 'data-soui-position', item);
        styleTest(popover, positionStyleMap[item]);
      });
    });
  });
  test('should render that is controll', () => {
    render(<PopoverControll />);
    screen.debug();
  });
});
