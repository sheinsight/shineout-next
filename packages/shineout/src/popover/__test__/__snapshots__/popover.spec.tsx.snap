// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover[Base] render snapshot default 1`] = `
<div
  class="popover-wrapper-0-2-65 popover-wrapperOpen-0-2-66"
  data-soui-position="bottom"
  style="position: absolute; z-index: 1060; left: 0px; transform: translateX(-50%); top: 0px;"
>
  <div
    class="popover-arrow-0-2-67"
  />
  <div
    class="popover-content-0-2-68 popover-text-0-2-69"
  >
    some Text
  </div>
</div>
`;

exports[`Popover[Base] should render default 1`] = `null`;

exports[`Popover[Base] should render when set className and style 1`] = `null`;

exports[`Popover[Snapshot] render snapshot default 1`] = `null`;
